Adding UpdateSchemaOptions in Upload and Query

amanda-tarafa opened the pull request.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/UploadTest.cs:
```diff
@@ -134,8 +190,8 @@ public void UploadAvro_IgnoresSchema()
             Assert.Equal(1, rows.Count);
-            Assert.Equal(100, (long) rows[0]["re"]);
-            Assert.Equal(200, (long) rows[0]["im"]);
+            Assert.Equal(100, (long)rows[0]["re"]);
```
I'm not sure what's been automatically formatting the casts, but it would be good if you could turn it off. (I don't much mind whether there's a space or not, but it's better not to have unrelated changes in the commit.)

Don't worry about undoing these ones - let's just focus on avoiding them in the future.

amanda-tarafa replied:
That would be Edit -> Advanced -> Format Document
Won't do it again.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/UploadTest.cs:
```diff
@@ -183,8 +239,52 @@ public void UploadAvro_IgnoresSchema()
+                    {
+                        WriteDisposition = WriteDisposition.WriteAppend,
+                        DestinationSchemaUpdateOptions = SchemaUpdateOption.AllowFieldAddition,
+
```
Nit - blank line.

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/EnumMapTest.cs:
```diff
@@ -36,13 +37,35 @@ public void InvalidConversion()
+            Assert.Throws<ArgumentException>(() => EnumMap<SchemaUpdateOption>.ToApiValues((SchemaUpdateOption)100));
+        }
+
+        [Theory]
```
InvalidFlagsConversion test?

amanda-tarafa replied:
I added it on the method above InvalidConversion since that was dealing with the two types of coversions  that existed till now, so added one Assert for the ToApiValues conversion.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/EnumMap.cs:
```diff
@@ -71,6 +74,41 @@ internal static string ToApiValue(T value, string paramName = "value")
+        /// of the flags set in <paramref name="value"/>.</returns>
+        internal static ISet<string> ToApiValues(T value)
+        {
+            Enum eValue = value as Enum;
```
I think we can do all of this without any calls to `Enum.GetValues` etc after the type initialization.
It's probably simplest if we get the code merged as-is, then I'll do a follow-up PR. (I know what I'm thinking of, but it's harder to describe than just to do :)

amanda-tarafa replied:
I think I understood what you described on the previous PR using an extra dictionary of int to string. This was just a faster change :).
Let me try your proposal and if I get it wrong you can do the follow up PR.

amanda-tarafa commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/Google.Cloud.BigQuery.V2.IntegrationTests.csproj:
```diff
@@ -24,8 +24,14 @@
     <EmbeddedResource Include="*.parquet" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="after_schema_update.avro" />
```
No idea how that got there, I just used VS to add the files and set them as EmbeddedResource for build. I will edit the .csproj to remove that and see if nothing breaks.

amanda-tarafa commented:
@jskeet PTAL. I addressed your comments and changed `EnumMap` to what I understood you were proposing.
Also LMK if you prefer a specific `InvalidFlagsConversion` test method (see my answer to your comment).
Thanks!

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/EnumMap.cs:
```diff
@@ -71,6 +76,35 @@ internal static string ToApiValue(T value, string paramName = "value")
+        {
+            int intValue = (int)(object) value;
+            ISet<string> apiValues = new HashSet<string>();
+            foreach (int oneValue in s_intValueToString.Keys)
```
If you iterate over the key value pairs instead of just the keys, you don't need the lookup in the Add call.

DaveTCode commented:
I feel like I'm probably being daft but I don't think this covers load jobs as well as query jobs? http://googlecloudplatform.github.io/google-cloud-dotnet/docs/Google.Cloud.BigQuery.V2/history.html claims it does but looking at the code for CreateLoadJobOptions.cs I think that was missed (perhaps deliberately?) as part of this PR.

I'm happy to submit a PR for it but it's pretty much just copy paste coding from the work that @amanda-tarafa has already done so maybe not worth me doing?

amanda-tarafa commented:
@DaveTCode when we said that we were covering load jobs we were referring to all the Upload*Options.cs that hold options for uploading data in different formats to tables that might already exist.
Let me check if CreateLoadJobOptions.cs slipped past us and we didn't included changes there when we had to. I will update here once I've checked.

DaveTCode commented:
Great! Thanks :) I'd send a patch but frankly I just copy pasted your code around so I don't think there's any point!
